docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,205 @@
+
+#### Quick Primer on JWS and JWT
+
+[JSON Web Signatures](https://tools.ietf.org/html/rfc7515) are a way to sign, serialize and verify a payload.  It supports both symmetric keys (aka shared secrets) along with asymmetric keys (aka public key infrastructure or key pairs).  The JWS is split in to 3 parts:
+1. a header about how it is signed
+2. the clear text payload
+3. the signature.
+
+There are a couple of different ways of encoding this data -- either as a JSON object or as a set of BASE64URL strings for including in headers or URL parameters.  In this case, we are using a shared secret and the HMAC-SHA256 signing algorithm and encoding it as a JSON object.  The popular JWT (JSON Web Tokens) specification is a type of JWS.
+
+The JWS specification [describes how to encode](https://tools.ietf.org/html/rfc7515#appendix-F) "detached content".  In this way the signature is calculated as normal but the content isn't included in the signature.
+
+#### NEW: JWT authentication
+
+To avoid passing an unencoded token we will instead create a short lived JWT based on the token.  The JWT will be constructed as so:
+
+* **Header**
+  * **typ**: `JWT`. REQUIRED.
+  * **alg**: `HS256` for HMAC SHA-256. REQUIRED.
+  * **kid**: The token-id. REQUIRED.
+* **Claims**
+  * **sub**: The token-id. REQUIRED.
+  * **aud**: The long term globally unique cluster identifier.  This is not implemented yet so is OPTIONAL.
+  * **exp**: An expiration time for this JWT.  This must be no more than 2 minutes in the future. REQUIRED.
+  * **iat**: The time that the JWT was issued.  OPTIONAL.
+
+**Question:** The 2 minute value is picked to be short enough to discourage replay attacks without introducing the complexity of tracking nonces.  If we mandated the use of nonces, we'd have to track which nonces were used across HA instances.  Is there a better value than 2 minutes?  If we see clock skew greater or connectivity problems greater than 2 minutes we'll probably have other problems.
+
+#### NEW: `kube-public` namespace
+
+Kubernetes ConfigMaps are per-namespace and are generally only visible to principals that have read access on that namespace.  To create a config map that *everyone* can see, we introduce a new `kube-public` namespace.  This namespace, by convention, is readable by all authenticated users.
+
+#### NEW: JWS content signing
+
+When the JWT flow is used, we go one step further and sign the payload of the response using a JWS with detached content.
+
+We compute the JWS as normal (using the token provided in the JWT and `HS256`) and then remove the middle section that contains the BASE64URL encoded content.  This shortened JWS (which only has the JWS header and signature) is then put in a new HTTP header.  This header is `X-Kubernetes-JWS`.
+
+We use a "detached content" JWS to ease implementation.  This can be implemented a simple HTTP request filter without having to modify the subtle content negotiation mechanisms in the API server.
```

Comment:
Fully acknowledging that you all know this inside out, from the perspective of someone who doesn't, putting this into a config map, especially one with a key per token that have to be maintained across new/removed tokens and API servers, does not seem like a clean solution. 

Admittedly signing on the fly as a core request response type/filter is a bit dramatic if this is our only expected usage. (though is it in any way a negative to have available in the API? It's probably not a lot of code and could prove useful.)

If the answer to the above is no, I really think we should re-examine this as a core API and encapsulate the JWS logic there. To me, this does not feel like "configuration", it's querying live state of the cluster with respect to what API servers and tokens are in play. It should be queried live IMO, and I'd love for us to seek to use it in the discussions around in-cluster HA (cc @timothysc @pires), and maybe converge with the pkg/kubediscovery federation code we recently discovered. I know there are strong motivations to avoiding a core API for this which I do not fully understand yet, but we're making a meta-API here and very unusual cross-cutting concerns to maintain the same data signed with multiple keys to avoid a core API. With those two options on the table, this is not the one that feels best to me.

Sorry for rehashing the old argument just a perspective from someone trying to get to grips with this.